reference/reference.go

```diff
@@ -0,0 +1,269 @@
+//     repository                      := hostname ['/' component]*
+//     hostname                        := component [port]
+//     component                       := alpha-numeric [separator alpha-numeric]*
+//     alpha-numeric                   := /[a-z0-9]+/
+//     separator                       := /[._-]/
+//     port                            := /:[0-9]+/
+//     tag                             := /[\w][\w.-]{0,127}/
+//
+//     // from the digest package
+//     digest                          := digest-algorithm ":" digest-hex
+//     digest-algorithm                := digest-algorithm-component [digest-algorithm-separator digest-algorithm-component]
+//     digest-algorithm-separator      := /[+.-_]/
+//     digest-algorithm-component      := /[A-Za-z]/ /[A-Za-z0-9]*/
+//     digest-hex                      := /[A-Za-z0-9_-]+/ ; supports hex bytes or url safe base64
+package reference
+
+import (
+	"errors"
+	"regexp"
+
+	"github.com/docker/distribution/digest"
+)
+
+var (
+	tagRegexp    = regexp.MustCompile(`[\w][\w.-]{0,127}`)
+	alphaNumeric = regexp.MustCompile(`[a-z0-9]+`)
+	portRegexp   = regexp.MustCompile(`:[0-9]+`)
+)
+
+// ErrReferenceInvalidFormat represents an error while trying to parse a string as a reference.
+var ErrReferenceInvalidFormat = errors.New("invalid reference format")
+
+// Reference abstracts types that reference images in a certain way.
+type Reference interface {
+	// Returns the repository part of a reference
+	Repository() string
+	// Returns the entire reference, including the repository
+	String() string
+}
+
+type parser struct {
+	err error
+}
+
+// parseComponent parses a component of the repository.
+//
+// 	component                       := alpha-numeric [separator alpha-numeric]*
+// 	separator                       := /[._-]/
+func (p *parser) parseComponent(s string) (component, tail string) {
+	tail = s
+	if p.err != nil {
+		return
+	}
+	component = alphaNumeric.FindString(s)
+	if len(component) == 0 {
+		p.err = ErrReferenceInvalidFormat
+		return
+	}
+	tail = s[len(component):]
+	for len(tail) > 0 {
+		switch tail[0] {
+		// separator
+		case '.', '_', '-':
+			c := alphaNumeric.FindString(tail[1:])
+			if len(c) == 0 {
+				return
+			}
+			component += tail[:len(c)+1]
+			tail = tail[len(c)+1:]
+			continue
+		}
+		break
+	}
+	return
+}
+
+// parseHostname parses the hostname part of the repository.
+//
+//     hostname                        := component [port]
+func (p *parser) parseHostname(s string) (hostname, tail string) {
+	tail = s
+	if p.err != nil {
+		return
+	}
+	hostname, tail = p.parseComponent(s)
+	port := portRegexp.FindString(tail)
+	if len(port) > 0 {
+		hostname += port
+		tail = tail[len(port):]
+	}
+	return
+}
+
+// parseRepository parses the repository part of a reference.
+func (p *parser) parseRepository(s string) (repository, tail string) {
+	tail = s
+	if p.err != nil {
+		return
+	}
+	repository, tail = p.parseHostname(s)
```

Comment:
This function doesn't appear that it would handle a single component repository with a numeric tag correctly. The port should only be considered if there are remaining components.